lib/connections/limiter.go

```diff
@@ -12,53 +12,111 @@ import (
 	"sync/atomic"
 
 	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
 	"golang.org/x/net/context"
 	"golang.org/x/time/rate"
 )
 
+type deviceLimiters map[protocol.DeviceID]*rate.Limiter
+
 // limiter manages a read and write rate limit, reacting to config changes
 // as appropriate.
 type limiter struct {
-	write     *rate.Limiter
-	read      *rate.Limiter
-	limitsLAN atomicBool
+	write              *rate.Limiter
+	read               *rate.Limiter
+	limitsLAN          atomicBool
+	deviceReadLimiter  deviceLimiters
+	deviceWriteLimiter deviceLimiters
+	myID               protocol.DeviceID
 }
 
 const limiterBurstSize = 4 * 128 << 10
 
-func newLimiter(cfg *config.Wrapper) *limiter {
+func newLimiter(deviceID protocol.DeviceID, cfg *config.Wrapper) *limiter {
 	l := &limiter{
-		write: rate.NewLimiter(rate.Inf, limiterBurstSize),
-		read:  rate.NewLimiter(rate.Inf, limiterBurstSize),
+		write:              rate.NewLimiter(rate.Inf, limiterBurstSize),
+		read:               rate.NewLimiter(rate.Inf, limiterBurstSize),
+		deviceReadLimiter:  make(deviceLimiters),
+		deviceWriteLimiter: make(deviceLimiters),
+		myID:               deviceID,
+	}
+
+	// Keep read/write limiters for every connected device
+	devices := getInitialDevicesConfiguration(cfg.RawCopy())
+	for _, v := range devices {
+		l.deviceReadLimiter[v.DeviceID] = rate.NewLimiter(rate.Inf, limiterBurstSize)
+		l.deviceWriteLimiter[v.DeviceID] = rate.NewLimiter(rate.Inf, limiterBurstSize)
 	}
+
 	cfg.Subscribe(l)
-	prev := config.Configuration{Options: config.OptionsConfiguration{MaxRecvKbps: -1, MaxSendKbps: -1}}
+	prev := config.Configuration{Options: config.OptionsConfiguration{MaxRecvKbps: -1, MaxSendKbps: -1}, Devices: devices}
 	l.CommitConfiguration(prev, cfg.RawCopy())
 	return l
 }
 
-func (lim *limiter) newReadLimiter(r io.Reader, isLAN bool) io.Reader {
-	return &limitedReader{reader: r, limiter: lim, isLAN: isLAN}
+func getInitialDevicesConfiguration(cfgCopy config.Configuration) config.DeviceConfigurationList {
```

Comment:
This is only called once, I'd say that it's fine to wedge in the loop inline in the constructor.